File: packages/client-react/src/transport/socket.ts
import type { Event } from '@rfnry/chat-protocol'
import { toEvent } from '@rfnry/chat-protocol'
import { io, type Socket } from 'socket.io-client'
import { SocketTransportError } from '../errors'

export type SocketAuthPayload = {
  headers?: Record<string, string>
  auth?: Record<string, unknown>
}

export type SocketTransportOptions = {
  baseUrl: string
  socketPath?: string
  authenticate?: () => Promise<SocketAuthPayload>
  ackTimeoutMs?: number
  reconnectionAttempts?: number
  onReconnectFailed?: () => void
}

export class SocketTransport {
  readonly baseUrl: string
  readonly socketPath: string
  private readonly authenticate?: SocketTransportOptions['authenticate']
  private readonly ackTimeoutMs: number
  private readonly reconnectionAttempts: number
  private readonly onReconnectFailed?: () => void
  private socket: Socket | null = null

  constructor(opts: SocketTransportOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/$/, '')
    this.socketPath = opts.socketPath ?? '/chat/ws'
    this.authenticate = opts.authenticate
    this.ackTimeoutMs = opts.ackTimeoutMs ?? 15_000
    this.reconnectionAttempts = opts.reconnectionAttempts ?? 20
    this.onReconnectFailed = opts.onReconnectFailed
  }

  async connect(): Promise<void> {
    const auth = this.authenticate ? ((await this.authenticate()).auth ?? {}) : {}
    const socket = io(this.baseUrl, {
      path: this.socketPath,
      transports: ['websocket'],
      auth,
      reconnectionDelayMax: 30_000,
      reconnectionAttempts: this.reconnectionAttempts,
    })
    this.socket = socket

    if (this.onReconnectFailed) {
      socket.on('reconnect_failed', this.onReconnectFailed)
    }

    await new Promise<void>((resolve, reject) => {
      socket.once('connect', () => resolve())
      socket.once('connect_error', (err) => reject(err))
    })
  }

  async disconnect(): Promise<void> {
    if (this.socket) {
      this.socket.disconnect()
      this.socket = null
    }
  }

  on(event: string, handler: (data: unknown) => void): () => void {
    if (!this.socket) throw new Error('not connected')
    this.socket.on(event, handler)
    return () => this.socket?.off(event, handler)
  }

  async joinThread(
    threadId: string,
    since?: { createdAt: string; id: string }
  ): Promise<{ threadId: string; replayed: Event[]; replayTruncated: boolean }> {
    const payload: Record<string, unknown> = { thread_id: threadId }
    if (since) {
      payload.since = { created_at: since.createdAt, id: since.id }
    }
    const response = await this._emitWithAck<{
      thread_id?: string
      replayed?: unknown[]
      replay_truncated?: boolean
      error?: { code: string; message: string }
    }>('thread:join', payload)
    if (response.error) {
      throw new SocketTransportError(response.error.code, response.error.message)
    }
    return {
      threadId: response.thread_id ?? threadId,
      replayed: (response.replayed ?? []).map((e) => toEvent(e as never)),
      replayTruncated: response.replay_truncated ?? false,
    }
  }

  async leaveThread(threadId: string): Promise<void> {
    await this._emitWithAck('thread:leave', { thread_id: threadId })
  }

  async sendEvent(threadId: string, event: Record<string, unknown>): Promise<Event> {
    const reply = await this._emitWithAck<{
      event?: unknown
      error?: { code: string; message: string }
    }>('event:send', { thread_id: threadId, event })
    if (reply.error) {
      throw new SocketTransportError(reply.error.code, reply.error.message)
    }
    return toEvent(reply.event as never)
  }

  async beginRun(
    threadId: string,
    opts: { triggeredByEventId?: string; idempotencyKey?: string } = {}
  ): Promise<{ runId: string; status: string }> {
    const payload: Record<string, unknown> = { thread_id: threadId }
    if (opts.triggeredByEventId) payload.triggered_by_event_id = opts.triggeredByEventId
    if (opts.idempotencyKey) payload.idempotency_key = opts.idempotencyKey
    const reply = await this._emitWithAck<{
      run_id?: string
      status?: string
      error?: { code: string; message: string }
    }>('run:begin', payload)
    if (reply.error) {
      throw new SocketTransportError(reply.error.code, reply.error.message)
    }
    return { runId: reply.run_id!, status: reply.status! }
  }

  async endRun(
    runId: string,
    opts: { error?: { code: string; message: string } } = {}
  ): Promise<{ runId: string; status: string }> {
    const payload: Record<string, unknown> = { run_id: runId }
    if (opts.error) payload.error = opts.error
    const reply = await this._emitWithAck<{
      run_id?: string
      status?: string
      error?: { code: string; message: string }
    }>('run:end', payload)
    if (reply.error) {
      throw new SocketTransportError(reply.error.code, reply.error.message)
    }
    return { runId: reply.run_id!, status: reply.status! }
  }

  async sendStreamStart(frame: {
    eventId: string
    threadId: string
    runId: string
    targetType: 'message' | 'reasoning'
    author: Record<string, unknown>
  }): Promise<void> {
    await this._sendStreamFrame('stream:start', {
      event_id: frame.eventId,
      thread_id: frame.threadId,
      run_id: frame.runId,
      target_type: frame.targetType,
      author: frame.author,
    })
  }

  async sendStreamDelta(frame: { eventId: string; threadId: string; text: string }): Promise<void> {
    await this._sendStreamFrame('stream:delta', {
      event_id: frame.eventId,
      thread_id: frame.threadId,
      text: frame.text,
    })
  }

  async sendStreamEnd(frame: {
    eventId: string
    threadId: string
    error?: { code: string; message: string } | null
  }): Promise<void> {
    const payload: Record<string, unknown> = {
      event_id: frame.eventId,
      thread_id: frame.threadId,
    }
    if (frame.error) payload.error = frame.error
    await this._sendStreamFrame('stream:end', payload)
  }

  private async _emitWithAck<T>(event: string, payload: Record<string, unknown>): Promise<T> {
    if (!this.socket) throw new Error('not connected')
    return (await this.socket.timeout(this.ackTimeoutMs).emitWithAck(event, payload)) as T
  }

  private async _sendStreamFrame(event: string, payload: Record<string, unknown>): Promise<void> {
    const reply = await this._emitWithAck<{
      ok?: boolean
      error?: { code: string; message: string }
    }>(event, payload)
    if (reply.error) {
      throw new SocketTransportError(reply.error.code, reply.error.message)
    }
  }

  get rawSocket(): Socket | null {
    return this.socket
  }
}

File: packages/client-python/src/rfnry_chat_client/transport/socket.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any, Protocol

AuthenticatePayload = dict[str, Any]
AuthenticateCallable = Callable[[], Awaitable[AuthenticatePayload]]
RawEventHandler = Callable[[dict[str, Any]], Awaitable[None]]


class _SioClient(Protocol):
    async def connect(
        self,
        url: str,
        *,
        auth: dict[str, Any] | None = ...,
        headers: dict[str, str] | None = ...,
        transports: list[str] | None = ...,
        socketio_path: str | None = ...,
    ) -> None: ...
    async def disconnect(self) -> None: ...
    def on(self, event: str, handler: Any = ...) -> Any: ...
    async def emit(self, event: str, data: Any = ...) -> None: ...
    async def call(self, event: str, data: Any = ..., *, timeout: float | None = ...) -> Any: ...


class SocketTransportError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(f"{code}: {message}")
        self.code = code
        self.message = message


class SocketTransport:
    def __init__(
        self,
        *,
        base_url: str,
        sio_client: _SioClient | None = None,
        socketio_path: str = "/chat/ws",
        authenticate: AuthenticateCallable | None = None,
    ) -> None:
        if sio_client is None:
            import socketio

            sio_client = socketio.AsyncClient()
        self._sio: _SioClient = sio_client
        self._base_url = base_url.rstrip("/")
        self._socketio_path = socketio_path
        self._authenticate = authenticate

    @property
    def base_url(self) -> str:
        return self._base_url

    @property
    def socketio_path(self) -> str:
        return self._socketio_path

    async def connect(self) -> None:
        auth_payload: AuthenticatePayload = {}
        if self._authenticate is not None:
            auth_payload = await self._authenticate()
        await self._sio.connect(
            self._base_url,
            auth=auth_payload.get("auth"),
            headers=auth_payload.get("headers"),
            transports=["websocket"],
            socketio_path=self._socketio_path,
        )

    async def disconnect(self) -> None:
        await self._sio.disconnect()

    def on_raw_event(self, name: str, handler: RawEventHandler) -> None:
        self._sio.on(name, handler)

    async def join_thread(
        self,
        thread_id: str,
        since: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"thread_id": thread_id}
        if since is not None:
            payload["since"] = since
        reply = await self._sio.call("thread:join", payload)
        _raise_if_error(reply)
        return reply

    async def leave_thread(self, thread_id: str) -> dict[str, Any]:
        reply = await self._sio.call("thread:leave", {"thread_id": thread_id})
        _raise_if_error(reply)
        return reply

    async def send_message(self, thread_id: str, draft: dict[str, Any]) -> dict[str, Any]:
        reply = await self._sio.call("message:send", {"thread_id": thread_id, "draft": draft})
        _raise_if_error(reply)
        return reply

    async def send_event(self, thread_id: str, event: dict[str, Any]) -> dict[str, Any]:
        reply = await self._sio.call("event:send", {"thread_id": thread_id, "event": event})
        _raise_if_error(reply)
        return reply

    async def begin_run(
        self,
        thread_id: str,
        *,
        triggered_by_event_id: str | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"thread_id": thread_id}
        if triggered_by_event_id is not None:
            payload["triggered_by_event_id"] = triggered_by_event_id
        if idempotency_key is not None:
            payload["idempotency_key"] = idempotency_key
        reply = await self._sio.call("run:begin", payload)
        _raise_if_error(reply)
        return reply

    async def end_run(
        self,
        run_id: str,
        *,
        error: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"run_id": run_id}
        if error is not None:
            payload["error"] = error
        reply = await self._sio.call("run:end", payload)
        _raise_if_error(reply)
        return reply

    async def cancel_run(self, run_id: str) -> dict[str, Any]:
        reply = await self._sio.call("run:cancel", {"run_id": run_id})
        _raise_if_error(reply)
        return reply

    async def send_stream_start(self, frame: dict[str, Any]) -> dict[str, Any]:
        reply = await self._sio.call("stream:start", frame)
        _raise_if_error(reply)
        return reply

    async def send_stream_delta(self, frame: dict[str, Any]) -> None:
        # Fire-and-forget: token streams must not block on per-frame RTT.
        # Awaiting an ack for every delta caps throughput at ~1/RTT tokens/sec
        # (~200 tok/s at 5ms RTT). stream:start and stream:end keep using call
        # because they need ordering/error signaling.
        await self._sio.emit("stream:delta", frame)

    async def send_stream_end(self, frame: dict[str, Any]) -> dict[str, Any]:
        reply = await self._sio.call("stream:end", frame)
        _raise_if_error(reply)
        return reply


def _raise_if_error(reply: Any) -> None:
    if isinstance(reply, dict) and "error" in reply:
        err = reply["error"]
        raise SocketTransportError(err.get("code", "unknown"), err.get("message", ""))
